### PR TITLE
Do not create temporary object during [[Get]] operation with primitive base

### PR DESCRIPTION
**Pull request by rerobika**
JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

**Comment by rerobika** [comment]
Perf results:

Benchmark | Heap (bytes) | Stack (bytes) | Perf (sec) |
----: | ----: | ----: | ----: | 
3d-cube.js | 10672 -> 10672 : 0.000% | 14152 -> 14152 : 0.000% | 0.729 -> 0.732 : -0.377% | 
3d-morph.js | 288984 -> 288984 : 0.000% | 1464 -> 1464 : 0.000% | 0.714 -> 0.717 : -0.356% | 
3d-raytrace.js | 116424 -> 116424 : 0.000% | 2688 -> 2688 : 0.000% | 0.869 -> 0.870 : -0.136% | 
access-binary-trees.js | 27120 -> 27120 : 0.000% | 3328 -> 3328 : 0.000% | 0.451 -> 0.450 : +0.297% | 
access-fannkuch.js | 1560 -> 1560 : 0.000% | 1536 -> 1536 : 0.000% | 1.758 -> 1.750 : +0.452% | 
access-nbody.js | 4088 -> 4088 : 0.000% | 1552 -> 1552 : 0.000% | 0.989 -> 0.986 : +0.261% | 
bitops-3bit-bits-in-byte.js | 792 -> 792 : 0.000% | 1480 -> 1480 : 0.000% | 0.477 -> 0.477 : +0.053% | 
bitops-bits-in-byte.js | 712 -> 712 : 0.000% | 1480 -> 1480 : 0.000% | 0.648 -> 0.648 : +0.111% | 
bitops-bitwise-and.js | 536 -> 536 : 0.000% | 1320 -> 1320 : 0.000% | 0.807 -> 0.810 : -0.427% | 
bitops-nsieve-bits.js | 60904 -> 60904 : 0.000% | 1480 -> 1480 : 0.000% | 1.011 -> 1.009 : +0.254% | 
controlflow-recursive.js | 1000 -> 1000 : 0.000% | 63624 -> 63624 : 0.000% | 0.316 -> 0.321 : -1.361% | 
crypto-aes.js | 26896 -> 27832 : -3.480% | 2084 -> 2084 : 0.000% | 0.624 -> 0.619 : +0.888% | 
crypto-md5.js | 70648 -> 70496 : +0.215% | 1752 -> 1752 : 0.000% | 0.513 -> 0.497 : **+3.225%** | 
crypto-sha1.js | 45896 -> 45832 : +0.139% | 1608 -> 1608 : 0.000% | 0.521 -> 0.512 : **+1.910%** | 
date-format-tofte.js | 8664 -> 8664 : 0.000% | 1968 -> 1968 : 0.000% | 0.659 -> 0.658 : +0.207% | 
date-format-xparb.js | 15400 -> 15384 : +0.104% | 2528 -> 2416 : **+4.430%** | 0.485 -> 0.485 : +0.108% | 
math-cordic.js | 1696 -> 1696 : 0.000% | 1480 -> 1480 : 0.000% | 1.107 -> 1.108 : -0.030% | 
math-partial-sums.js | 1256 -> 1256 : 0.000% | 1456 -> 1456 : 0.000% | 0.632 -> 0.634 : -0.338% | 
math-spectral-norm.js | 2920 -> 2920 : 0.000% | 1480 -> 1480 : 0.000% | 0.468 -> 0.468 : -0.064% | 
string-base64.js | 89840 -> 89840 : 0.000% | 1528 -> 1528 : 0.000% | 1.188 -> 0.888 : **+25.246%** | 
string-fasta.js | 3552 -> 3552 : 0.000% | 1464 -> 1464 : 0.000% | 1.054 -> 1.039 : **+1.421%** | 
Geometric mean: | -0.141% | +0.216% | +1.667% | 

Binary (bytes) | master(cbae33a689) | patch(fdf93ba235) | Diff |
----: | ----: | ----: | ----: |
size | 132620 | 132620 | 0 bytes |
.rodata | 11746 | 11746 | 0 bytes |
.dynstr | 212 | 212 | 0 bytes |
.rel.plt | 160 | 160 | 0 bytes |
.interp | 25 | 25 | 0 bytes |
.dynsym | 416 | 416 | 0 bytes |
.gnu.hash | 212 | 212 | 0 bytes |
.text | 114712 | 114768 | +56 bytes |
.comment | 85 | 85 | 0 bytes |
.shstrtab | 226 | 226 | 0 bytes |
.data | 8 | 8 | 0 bytes |
.ARM.exidx | 8 | 8 | 0 bytes |
.rel.dyn | 16 | 16 | 0 bytes |
.init | 12 | 12 | 0 bytes |
.got | 96 | 96 | 0 bytes |
.plt | 264 | 264 | 0 bytes |
.note.ABI-tag | 32 | 32 | 0 bytes |
.gnu.version_r | 32 | 32 | 0 bytes |
.bss | 1575096 | 1575096 | 0 bytes |
.fini | 8 | 8 | 0 bytes |
.hash | 180 | 180 | 0 bytes |
.gnu.version | 52 | 52 | 0 bytes |
.fini_array | 4 | 4 | 0 bytes |
.init_array | 4 | 4 | 0 bytes |
.dynamic | 240 | 240 | 0 bytes |
.eh_frame | 4 | 4 | 0 bytes |
.ARM.attributes | 53 | 53 | 0 bytes |

File: jerry-core/ecma/operations/ecma-get-put-value.c [hunk 1]
```diff
@@ -120,59 +120,63 @@ ecma_value_t
...
+      return ecma_make_uint32_value (ecma_string_get_length (string_p));
     }
 
-    if (object_p->u2.prototype_cp == JMEM_CP_NULL)
+    uint32_t index = ecma_string_get_array_index (property_name_p);
+
+    if (index != ECMA_STRING_NOT_ARRAY_INDEX)
     {
-      break;
+      if (index < ecma_string_get_length (string_p))
```
Comment (galpeter) [inline jerry-core/ecma/operations/ecma-get-put-value.c]: Would it be possible to merge the two if-s?

File: jerry-core/ecma/operations/ecma-get-put-value.c [hunk 2]
```diff
@@ -120,59 +120,63 @@ ecma_value_t
...
+      {
+        ecma_char_t char_at_idx = ecma_string_get_char_at_pos (string_p, index);
+        return ecma_make_string_value (ecma_new_ecma_string_from_code_unit (char_at_idx));
+      }
     }
 
-    object_p = ECMA_GET_NON_NULL_POINTER (ecma_object_t, object_p->u2.prototype_cp);
+#if ENABLED (JERRY_BUILTIN_STRING)
```
[the remark this answers is not in the document]
Comment (rerobika) [reply jerry-core/ecma/operations/ecma-get-put-value.c]: Due to [line_153](https://github.com/jerryscript-project/jerryscript/pull/3278/files#diff-0c6afc3abd8de3c7835e7ca20641b4d5R153) it cannot be performed.

**Comment by rerobika** [comment]
@galpeter Thanks for the review, I've updated the PR.

**Pull request merged**